CNV-40785: Configure guest for real-time

orelmisan opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->
Add the instructions to configure a RHEL 9 VM for real-time usage.

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s):
<!--- Specify the version or versions of OpenShift your PR applies to. -->
4.16+

Issue:
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->
[CNV-40785](https://issues.redhat.com/browse/CNV-40785)

Link to docs preview:
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->
https://75714--ocpdocs-pr.netlify.app/openshift-enterprise/latest/virt/virtual_machines/advanced_vm_management/virt-configuring-cluster-realtime-workloads.html#virt-configuring-vm-real-time_virt-configuring-cluster-realtime-workloads

QE review:
- [X] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

orelmisan commented:
/test all

orelmisan commented:
/cc @RamLavi @segevyos @matosatti @yairpod

ngu-niny commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -67,38 +67,63 @@ spec:
 ----
-$ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf
+$ echo isolated_cores=2-3 > /etc/tuned/realtime-virtual-guest-variables.conf
+$ echo isolate_managed_irq=Y >> /etc/tuned/realtime-virtual-guest-variables.conf
```
Usually this setting has already been in the conf file, maybe it's better to have a check before setting it.

ngu-niny commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -67,38 +67,63 @@ spec:
 [source,terminal]
 ----
-$ echo isolated_cores=2-9 > /etc/tuned/cpu-partitioning-variables.conf
+$ echo isolated_cores=2-3 > /etc/tuned/realtime-virtual-guest-variables.conf
```
Maybe a typo, in following cyclictest command, '-a 2-4' is used, while the isolated cores are 2-3 here.

orelmisan replied:
Thank you, changed to `2-3`.

orelmisan commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -67,38 +67,63 @@ spec:
 [source,terminal]
 ----
-$ dnf install -y tuned-profiles-cpu-partitioning
+$ dnf install kernel-rt tuned cloud-init rt-tests
```
Thank you, changed to `realtime-tests`.

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -67,38 +67,61 @@ spec:
 
 . Restart the VM to apply the changes.
 
+. Test the guest latencies
```
I would split this part into a separate `Verification` section

```suggestion
.Verification

. Use the `cyclictest` tool to verify that the real-time guest is configured properly:
```

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -67,38 +67,61 @@ spec:
+$ cyclictest --priority 1 --policy fifo -h 50 -a 2-3 --mainaffinity 0,1 -t 4 -m -q -i 200 -D 12h
+----
+
+After the test finishes, we can get the “Max Latencies” in the output, it should be less than 40 micro seconds.
```
Is there an example output that we can add here?

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -1,17 +1,13 @@
 // Module included in the following assemblies:
 //
 // * virt/virtual_machines/advanced_vm_management/virt-configuring-cluster-realtime-workloads.adoc
-
```
Need to add the empty line back

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -1,17 +1,13 @@
 :_mod-docs-content-type: PROCEDURE
 [id="virt-configuring-vm-real-time_{context}"]
 = Configuring a virtual machine for real-time workloads
-
```
Same here (add empty line)

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -1,17 +1,13 @@
-
-You can configure a virtual machines (VM) to run real-time workloads.
-
+You can configure a virtual machine (VM) to run real-time workloads.
```
Need empty line below

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -1,17 +1,13 @@
 .Prerequisites
 * Your cluster is configured to run real-time workloads.
 * You have installed the `virtctl` tool.
-
```
Add back the empty line before the `.Procedure` heading

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +45,126 @@ spec:
+----
+.. Edit `/etc/tuned/realtime-virtual-guest-variables.conf`:
++
+[source,terminal]
```
```suggestion
[source,conf]
```

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +45,126 @@ spec:
+$ grubby --args="iommu=pt intel_iommu=on default_hugepagesz=1G idle=poll" --update-kernel=$(grubby --default-kernel)
+----
+. Restart the VM to apply the changes.
+.Verification
```
Add empty line above the `.Verification` heading

sjhala-ccs reviewed (commented):
Suggestions to fix the spacing/indentation issues

yossisegev commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
-            pageSize: 1Gi <6>
+            pageSize: 1Gi # <6>
         terminationGracePeriodSeconds: 0
 # ...
```
As discussed - we need a more explicit VM manifest example, that perfectly suits realtime. It must include the `image` entry, because a realtime VM is not a standard one, and the user might not have created and pushed their own image.

orelmisan replied:
@phoracek @sjhala-ccs WDYT?

orelmisan replied:
BTW, the example DPDK VM spec does not have a disk as well https://docs.openshift.com/container-platform/4.15/virt/vm_networking/virt-using-dpdk-with-sriov.html#virt-configuring-vm-dpdk_virt-using-dpdk-with-sriov

phoracek replied:
This is not the only place where we leave out part of the VM yaml. I think it makes sense to document only the "special" parts, to make this easier to maintaine by making if future-proof, and also put the spotlight on the important attributes.

That being said, I will not fight against having a full example either.

orelmisan commented:
@yossisegev @ngu-niny Can we move forward with this PR?

yossisegev commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
++
+[source,terminal]
+----
+$ tuned-adm profile realtime-virtual-guest
```
Also requires root.

yossisegev commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
++
+[source,terminal]
+----
+$ grubby --set-default=<installed_rt_kernel_version>
```
1. I'm using the U/S image `quay.io/kiagnose/kubevirt-realtime-checkup-vm:v0.2.0`, which produces a CentOS VM.
The kernel version as retrieved by either `uname -r` or `hostnamectl` is invalid for this command:
```
[cloud-user@realtime-vm tuned]$ uname -r
4.18.0-348.rt7.130.el8.x86_64
[cloud-user@realtime-vm tuned]$ sudo grubby --set-default=4.18.0-348.rt7.130.el8.x86_64
The param 4.18.0-348.rt7.130.el8.x86_64 is incorrect
```

[UPDATE] The cyclictest passed successfully (i.e. `Max Latencies` < 40us) despite this, so @orelmisan @RamLavi - is this setting necessary?

2. Also requires root

orelmisan replied:
`quay.io/kiagnose/kubevirt-realtime-checkup-vm:v0.2.0` is based on CentOS stream 8 and the real-time checkup does additional configuration to it after it boots.
Please test it with a "clean" RHEL 9 installation.

yossisegev commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
++
+[source,terminal]
+----
+$ grubby --args="iommu=pt intel_iommu=on default_hugepagesz=1G idle=poll" --update-kernel=$(grubby --default-kernel)
```
root

yossisegev commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
++
+[source,terminal]
+----
+$ cyclictest --priority 1 --policy fifo -h 50 -a 2-3 --mainaffinity 0,1 -t 2 -m -q -i 200 -D 12h
```
1. The user should be alerted that the `-D` parameter determines the test duration, so if they use this example it means a 12h test.
2. This also requires root

orelmisan replied:
@sjhala-ccs can we somehow refer to the man page of cyclictest?

yossisegev replied:
> The user should be alerted that the -D parameter determines the test duration, so if they use this example it means a 12h test.

This should still be addressed.

orelmisan commented:
[Change](https://github.com/openshift/openshift-docs/compare/faa06412614c857285ee7ffde6c6c09f20864265..73374dbe61404d5123daf1cc9759ffbc09d2d198): Addressed @yossisegev's comments:
1. Added the `running` field to the VM spec.
2. Changed the shell symbol to a pound `#` in order to emphasize that all the commands inside the guest should be done as the root user.

orelmisan commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +49,127 @@ spec:
 [source,terminal]
 ----
-$ dnf install -y tuned-profiles-cpu-partitioning
+$ dnf install kernel-rt tuned cloud-init realtime-tests
```
Have you tested it with RHEL 9?
Please see @ngu-niny' [comment](https://github.com/openshift/openshift-docs/pull/75714#discussion_r1597832897).

yossisegev replied:
Following this comment, I tried a RHEL9 image which I took from `registry.redhat.io/rhel9/rhel-guest-image`.
The result was 
```No match for argument: kernel-rt
Error: Unable to find a match: kernel-rt
```
(the other packages were either installed successfully, or were already present).
So I tried following the instructions in https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/9/html-single/installing_rhel_9_for_real_time/index (after registering to subscription-manager), and the result was
```
$ sudo subscription-manager repos --enable rhel-9-for-x86_64-rt-rpms
Error: 'rhel-9-for-x86_64-rt-rpms' does not match a valid repository ID. Use "subscription-manager repos --list" to see valid repositories.
```
Maybe I'm underestimating the advanced user who is supposed to be the audience for this feature, but I think there are too many obstacles on the way for running a successful realtime check. Maybe we should re-consider the importance of providing a RHEL 9 VM image.
@orelmisan @phoracek WDYT?

ngu-niny replied:
I used the KVM Guest Image in https://access.redhat.com/downloads/content/479/ver=/rhel---9/9.4/x86_64/product-software in the VM manifest, as following url part shows.

spec:
  dataVolumeTemplates:
  - apiVersion: cdi.kubevirt.io/v1beta1
    kind: DataVolume
    metadata:
      creationTimestamp: null
      name: rhel94-rt-5cpu-84-1
    spec:
      source:
        http:
          url: https://access.cdn.redhat.com/content/origin/files/sha256/d3/d362e72518a2d7415d850b8177c814f6fd87f42ca1640bda17e98855e1d6baad/rhel-9.4-x86_64-kvm.qcow2?user=3124af3237f60c89937f85f9f98896b7&_auth_=1716880954_459335f349cede3ac1e894c9c9b9acb5

Then in the guest, do following steps to enable the corresponding repos.
# subscription-manager register
# subscription-manager repos --enable rhel-9-for-x86_64-rt-rpms
# subscription-manager repos --enable rhel-9-for-x86_64-nfv-rpms

It works for most of the steps except following two commands:
# dnf --enablerepo=rt install -y tuned-profiles-realtime
# dnf --enablerepo=nfv install -y tuned-profiles-nfv-guest

After delete the '--enablerepo=*', it works:
# dnf install -y tuned-profiles-realtime
# dnf install -y tuned-profiles-nfv-guest

yossisegev replied:
Eventually, Orel supplied a VM with DataVolume that managed to run successfully, apply all the additional commands and run the checkup successfully.
Thank you @orelmisan , and thank you as well @ngu-niny for your support here.

orelmisan commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -26,9 +26,9 @@ spec:
   template:
```
Done.

yossisegev commented:
> [Change](https://github.com/openshift/openshift-docs/compare/faa06412614c857285ee7ffde6c6c09f20864265..73374dbe61404d5123daf1cc9759ffbc09d2d198): > 2. Changed the shell symbol to a pound `#` in order to emphasize that all the commands inside the guest should be done as the root user.

I'm not sure that the pound prompt sign is enough for clarifying that root privileges are required.

yossisegev commented:
Except for addressing the `-D` (duration) parameter when running `cyclictest` - everything else looks good from my side.

orelmisan commented:
[Change](https://github.com/openshift/openshift-docs/compare/4e6caffb9354e7a4660de8819d64ec97be496101..742a5ab04f2529d817b4076c83d07447bae680dc): Added a partial explanation to the cyclictest command.

yossisegev commented:
/lgtm

ngu-niny commented:
LGTM.

Acked-by: Nini Gu <ngu@redhat.com>

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
+# netdev_queue_count=4
+----
+<1> The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
+<2> isolate_managed_irq should be set to `Y`
```
```suggestion
<2> Uncomment the `isolate_managed_irq=`Y` line to move kernel-managed interrupt requests out of isolated cores.
```

orelmisan replied:
This line is not commented by default.
Is there an alternative phrasing we could use?

sjhala-ccs replied:
Maybe change it to:
```suggestion
<2> Set the `isolate_managed_irq` parameter to `Y` to move kernel-managed interrupt requests out of isolated cores.
```

orelmisan replied:
Done.

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
+#
+# netdev_queue_count=4
+----
+<1> The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
```
```suggestion
<1> Set the first two CPUs (0 and 1) aside for house keeping tasks and isolate the rest for the real-time application.
```

orelmisan replied:
We actually set the isolated CPU, and the reserved are the ones left.
Is there an alternative phrasing you think we can use here?

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
 ----
-
-.. To achieve low-latency tuning by using the `cpu-partitioning` profile in the TuneD application, run the following commands:
+.. Enable the realtime and nvf repos:
```
```suggestion
.. Enable the real-time and network function virtualization (NFV) repositories:
```

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
-
-. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 8 operating system:
+. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 9 operating system:
 .. Run the following command to connect to the VM console:
```
```suggestion
.. Run the following command to connect to the VM console as a root user:
```

orelmisan replied:
This is not required (and is a potential security risk), since the command is run on the user's laptop for example.
The commands which are executed inside the guest, do required root privilages.

sjhala-ccs replied:
Understood. Then we can keep the wording and the command as is. I will mark the two comments related to this step as resolved.

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
 +
 [source,terminal]
 ----
 $ virtctl console <vm_name>
```
```suggestion
$ sudo virtctl console <vm_name>
```

orelmisan replied:
This is not required.
Please see https://github.com/openshift/openshift-docs/pull/75714#discussion_r1619266559

sjhala-ccs commented on modules/virt-configuring-vm-real-time.adoc:
```diff
@@ -49,56 +50,144 @@ spec:
-
-.. Configure huge pages by using the GRUB boot loader command-line interface. In the following example, 8 1G huge pages are specified.
++
+[NOTE]
```
If the suggested changes on lines 70 and 74 are acceptable, then we can remove this note.

orelmisan replied:
This note is required.
Is the phrasing correct?

sjhala-ccs replied:
The note placement seems weird. Is it possible to add an actual step instead instead of an admonition? Something like:
`.. Log in to the guest as a root user by using the following command:`

orelmisan replied:
There are at least thee routes a user could take:
1. Login directly as root (this is somethings disabled).
2. Login as a regular user and switching to root (using `su` or `sudo su`).
3. Login as a regular user and executing each command using `sudo`.

This is why I think we should not limit the instructions to use a single command.
WDYT?

sjhala-ccs replied:
For the purpose of this procedure, wouldn't option 2 make the most sense? Especially because all the commands after `virtctl console <vm_name>` have the `#` prompt which indicates a root user. How about `.. If you are not already logged in as a root user, switch to the root user account to execute the remaining configuration steps.`

I am not completely opposed to a Note, but we would need to change the indentation to left so that it doesn't seem like it applies to the previous step (visually). Adjusting the indentation might cause the list numbering to break.

orelmisan replied:
Done.

orelmisan commented:
[Change](https://github.com/openshift/openshift-docs/compare/742a5ab04f2529d817b4076c83d07447bae680dc..fd46175eba89b6587a0786e473103c7f26893125): Answered @sjhala-ccs's comments.

sjhala-ccs reviewed (commented):
LGTM

sjhala-ccs commented:
/cherrypick enterprise-4.16
